pl-faded-parsons.py

```diff
@@ -1,314 +1,577 @@
+            raise FileNotFoundError(
+                "\n"
+                f"\tCorrect answer not found at `{self._solution_path}`! \n"
+                '\tProvide an answer or set "showCorrectAnswer" to false in `./info.json`'
+            )
+
+        return self._solution_path
+
+    @property
+    def solution(self) -> str:
+        with open(self.solution_path, "r") as f:
+            return f.read()
+
+    def __init__(self, element_html: str, data: pl.QuestionData):
+        def get_child_text_by_tag(element, tag: str) -> str:
+            """get the innerHTML of the first child of `element` that has the tag `tag`
+            default value is empty string"""
+            return next((elem.text for elem in element if elem.tag == tag), "")
+
+        element: xml.HtmlElement = xml.fragment_fromstring(element_html)
+        self._element: xml.HtmlElement = element
+        pl.check_attribs(
+            element,
+            required_attribs=[
+                "answers-name",
+            ],
+            optional_attribs=[
+                "solution-path",
+                "format",
+                "language",
+                "file-name",
+                "solution-path",
+            ],
+        )
+
+        self.answers_name = pl.get_string_attrib(element, "answers-name")
+        self.format = FadedParsonsProblem.Format(
+            pl.get_string_attrib(element, "format", "right").replace("-", "_")
+        )
+        self.pre_text = get_child_text_by_tag(element, "pre-text").strip("\n")
+        self.post_text = get_child_text_by_tag(element, "post-text").strip("\n")
+        self.language: str = pl.get_string_attrib(element, "language", "")
+        self.out_filename = pl.get_string_attrib(element, "file-name", "user_code.py")
+        self.size = (
+            "narrow" if self.format == FadedParsonsProblem.Format.RIGHT else "wide"
+        )
+
+        self.markup = get_child_text_by_tag(self._element, "code-lines")
+        if not self.markup:
+            try:
+                path = os.path.join(
+                    self._options["question_path"],
+                    "serverFilesQuestion",
+                    "code_lines.txt",
+                )
+                with open(path, "r") as f:
+                    self.markup = f.read()
+            except:
+                self.markup = str(self._element.text)
+
+        if self.format == FadedParsonsProblem.Format.RIGHT and (
+            self.pre_text or self.post_text
+        ):
+            raise Exception(
+                "pre-text and post-text are not supported in right (horizontal) mode. "
+                + 'Add/set `format="bottom"` or `format="no-code"` to your element to use this feature.'
+            )
+
+        self._solution_path = pl.get_string_attrib(
+            element, "solution-path", "./solution"
+        )
+        self._max_distractors = 10  # this was hardcoded before
+        self._raw_answers = data["raw_submitted_answers"]
```

Comment:
Before, all entries in `data['raw_submitted_answers']` were prefixed with `answers_name`. Because we are using Submission here, I don't know what to do. I have yet to test it...

Reply:
`Submission` is just an abstraction to interact with it. We need to load the element _into_ a `Submission` instance. I'll add that in now